sql: Make `tree.ParseJSON` faster

miretskiy opened the pull request:
Switch `tree.ParseJSON` to use `json-iterator` library to convert input string directly into `tree.JSON`, without extra conversion to go native type.

The change is a drop in replacement for standard `encoding/json` library.  Allmost all metrics (speed per op, throughput, allocations/op) are improved significantly.  The new implementation tends to allocate more when encoding very large strings (256KB or larger).

This change should help for any insert query that converts strings to JSON -- including, large scale imports of JSON data.

Release note (performance improvement): Signficantly speed up conversion of strings to JSONb.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89117)
<!-- Reviewable:end -->

miretskiy commented:
Benchmark results are as follows:
```
name                  old time/op    new time/op     delta
ParseJSON/64-10         2.13µs ± 4%     0.92µs ± 2%   -56.85%  (p=0.000 n=24+25)
ParseJSON/256-10        2.32µs ± 1%     1.06µs ± 3%   -54.15%  (p=0.000 n=23+25)
ParseJSON/1024-10       3.07µs ± 2%     1.51µs ± 4%   -50.76%  (p=0.000 n=25+25)
ParseJSON/4096-10       6.05µs ± 5%     3.15µs ± 5%   -47.99%  (p=0.000 n=25+23)
ParseJSON/32768-10      32.4µs ± 7%     17.8µs ± 6%   -44.94%  (p=0.000 n=24+24)
ParseJSON/65536-10      61.7µs ± 5%     33.4µs ± 8%   -45.78%  (p=0.000 n=25+25)
ParseJSON/131072-10      120µs ± 6%       64µs ± 9%   -46.79%  (p=0.000 n=25+25)
ParseJSON/262144-10      237µs ± 7%      124µs ± 6%   -47.39%  (p=0.000 n=24+24)
ParseJSON/1048576-10     928µs ±15%      489µs ±10%   -47.37%  (p=0.000 n=25+25)

name                  old speed      new speed       delta
ParseJSON/64-10       32.9MB/s ± 3%   76.0MB/s ± 1%  +130.66%  (p=0.000 n=23+25)
ParseJSON/256-10      44.7MB/s ± 2%   98.4MB/s ± 2%  +119.98%  (p=0.000 n=24+24)
ParseJSON/1024-10     77.3MB/s ± 3%  159.1MB/s ± 3%  +105.99%  (p=0.000 n=24+25)
ParseJSON/4096-10      129MB/s ± 2%    250MB/s ± 2%   +93.51%  (p=0.000 n=23+25)
ParseJSON/32768-10     183MB/s ± 1%    332MB/s ± 0%   +81.04%  (p=0.000 n=24+23)
ParseJSON/65536-10     191MB/s ± 1%    352MB/s ± 2%   +84.45%  (p=0.000 n=25+23)
ParseJSON/131072-10    195MB/s ± 1%    368MB/s ± 1%   +89.11%  (p=0.000 n=21+25)
ParseJSON/262144-10    198MB/s ± 1%    375MB/s ± 1%   +89.52%  (p=0.000 n=25+23)
ParseJSON/1048576-10   202MB/s ± 4%    383MB/s ± 1%   +89.22%  (p=0.000 n=25+24)

name                  old alloc/op   new alloc/op    delta
ParseJSON/64-10         2.13kB ± 1%     1.07kB ± 2%   -49.90%  (p=0.000 n=25+25)
ParseJSON/256-10        2.21kB ± 1%     1.23kB ± 2%   -44.59%  (p=0.000 n=24+25)
ParseJSON/1024-10       2.81kB ± 2%     1.94kB ± 3%   -30.73%  (p=0.000 n=25+25)
ParseJSON/4096-10       5.68kB ± 6%     4.88kB ± 6%   -14.05%  (p=0.000 n=25+23)
ParseJSON/32768-10      32.6kB ± 9%     40.0kB ± 6%   +22.70%  (p=0.000 n=25+24)
ParseJSON/65536-10      62.6kB ± 6%     76.1kB ± 7%   +21.44%  (p=0.000 n=25+25)
ParseJSON/131072-10      122kB ± 6%      148kB ± 9%   +21.34%  (p=0.000 n=25+25)
ParseJSON/262144-10      239kB ± 6%      307kB ± 5%   +28.13%  (p=0.000 n=24+22)
ParseJSON/1048576-10     952kB ±13%     1312kB ± 9%   +37.83%  (p=0.000 n=25+25)

name                  old allocs/op  new allocs/op   delta
ParseJSON/64-10           34.0 ± 0%       18.0 ± 0%   -47.06%  (p=0.000 n=25+22)
ParseJSON/256-10          34.0 ± 0%       19.0 ± 0%   -44.12%  (p=0.000 n=24+24)
ParseJSON/1024-10         34.0 ± 0%       20.0 ± 0%   -41.18%  (p=0.000 n=24+23)
ParseJSON/4096-10         34.0 ± 0%       20.4 ± 3%   -40.12%  (p=0.000 n=20+25)
ParseJSON/32768-10        35.0 ± 0%       22.0 ± 0%   -37.14%  (p=0.000 n=23+20)
ParseJSON/65536-10        35.0 ± 0%       23.0 ± 0%   -34.29%  (p=0.000 n=21+21)
ParseJSON/131072-10       35.5 ± 1%       23.0 ± 0%   -35.17%  (p=0.000 n=25+21)
ParseJSON/262144-10       35.7 ± 2%       24.0 ± 0%   -32.81%  (p=0.000 n=25+24)
ParseJSON/1048576-10      36.0 ± 0%       25.4 ± 2%   -29.33%  (p=0.000 n=16+25)
```

Commit:
sql: Make `tree.ParseJSON` faster

Switch `tree.ParseJSON` to use `json-iterator` library to convert input
string directly into `tree.JSON`, without extra conversion to go native
type.

The change is a drop in replacement for standard `encoding/json`
library.  Allmost all metrics (speed per op, throughput, allocations/op)
are improved significantly.  The new implementation tends to allocate
more when encoding very large strings (256KB or larger).

This change should help for any insert query that converts strings to
JSON -- including, large scale imports of JSON data.

Release note (performance improvement): Signficantly speed up
conversion of strings to JSONb.

yuzefovich reviewed (approved):
Wow, nice speedups! Some nits, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-:-NDFKllj6MbNR6k17uDS:b1ijh5s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @miretskiy)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFHO5iD1t7yEE8Rdvi):*
nit: does this only affect the ccl features? Before reading the code, based on the comment, I'd expect either `sql change` or `performance improvement` category.
___
*[`pkg/util/json/json.go` line 860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFI4LaBsrV2k_CU0nc:-NDFI4LaBsrV2k_CU0nd:bta7t6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/util/json/json.go#L860)):*
> ```Go
> 	//  * simdjson-go: too restrictive; doesn't work on all platforms,
> 	//    does not parse raw json literals (true/false, etc); and is not
> 	//    a drop in replacement. easyjson,
> ```

nit: "easyjson" seems out of place.
___
*[`pkg/util/json/json.go` line 872 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFIB5ECwqCgWFz1xkt:-NDFIB5ECwqCgWFz1xku:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/util/json/json.go#L872)):*
> ```Go
> }
> 
> // parseJSONGoStd parses json using encoding/json library
> ```

nit: missing period.
___
*[`pkg/util/json/json.go` line 926 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFIsCJ30Xa5Xm2GHUz:-NDFIsCJ30Xa5Xm2GHV-:bbo9squ) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/util/json/json.go#L926)):*
> ```Go
> 		return FromBool(it.ReadBool())
> 	case jsoniter.ArrayValue:
> 		b := NewArrayBuilder(8)
> ```

nit: where does `8` come from? Maybe extract it in a local constant shared with `ObjectValue` case with a quick comment on how it was determined? (I'm assuming that the json iterator doesn't give us any info about the number of objects, right?)
___
*[`pkg/util/json/random.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFJdIn1PqGkW62yToi:-NDFJdIn1PqGkW62yToj:b-frf1o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/util/json/random.go#L55)):*
> ```Go
> }
> 
> // WithComplexity returns an option to set maximum complexity of JSON objection.
> ```

nit: probably `s/objection/objects/`.
___
*[`pkg/sql/sem/tree/json_test.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFKKXPErkxCc489Mxn:-NDFKKXQ6wVkITJeyctZ:b-7tp66p) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/sql/sem/tree/json_test.go#L85)):*
> ```Go
> 			j, err := jsonb.RandGen(rng, opts...)
> 			if err != nil {
> 				panic(err)
> ```

nit: `s/panic/b.Fatal/`
___
*[`pkg/sql/sem/tree/json_test.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFKUpJ2Lh6cVZvMItZ:-NDFKUpK3D8vjOzMj-Mc:bykm8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/sql/sem/tree/json_test.go#L98)):*
> ```Go
> 		b.Run(strconv.Itoa(maxStrLen), func(b *testing.B) {
> 			// Create test data for the specified string length.
> 			// Use the same test data for each implementation.
> ```

This comment hints at the desire to benchmark both implementations - why don't we do that?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89117)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @miretskiy)

___
*[`pkg/util/json/json.go` line 865 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFKxE76t0MK3Vn6mAU:-NDFKxE76t0MK3Vn6mAV:b-dl126y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/util/json/json.go#L865)):*
> ```Go
> )
> 
> var parseJSONImpl = UseJSONIter
> ```

nit: we could make the choice of the implementation a "metamorphic" variable with JSON Iter being the default.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89117)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFHO5iD1t7yEE8Rdvi):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: does this only affect the ccl features? Before reading the code, based on the comment, I'd expect either `sql change` or `performance improvement` category.
</blockquote></details>

Sorry, force of habit.
___
*[`pkg/util/json/json.go` line 865 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFKxE76t0MK3Vn6mAU:-NDFUmjN76PDVOKaz1dP:bi0qtfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/util/json/json.go#L865)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we could make the choice of the implementation a "metamorphic" variable with JSON Iter being the default.
</blockquote></details>

Great idea
___
*[`pkg/util/json/json.go` line 926 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFIsCJ30Xa5Xm2GHUz:-NDFVRa90zEiuyPYStP_:b799lx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/util/json/json.go#L926)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: where does `8` come from? Maybe extract it in a local constant shared with `ObjectValue` case with a quick comment on how it was determined? (I'm assuming that the json iterator doesn't give us any info about the number of objects, right?)
</blockquote></details>

Just randomly chosen because of the reasons you mention. Constant, with default value of 4, added 
___
*[`pkg/util/json/random.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFJdIn1PqGkW62yToi:-NDFVpzU8ZL45nDmRhN0:b2nhg5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/util/json/random.go#L55)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: probably `s/objection/objects/`.
</blockquote></details>

Ha!
___
*[`pkg/sql/sem/tree/json_test.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFKUpJ2Lh6cVZvMItZ:-NDFW-XRA4IAHNOOy3h9:bxxxt8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1522263b0f7db38597ab825c4f6f4d6fb3ae1e0/pkg/sql/sem/tree/json_test.go#L98)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This comment hints at the desire to benchmark both implementations - why don't we do that?
</blockquote></details>

Yeah, sorry. Wanted to run bench stat, removed code to test both, and forgot to put back 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89117)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @miretskiy)

___
*[`pkg/util/json/random.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFZMnX94cIK2noT3_J:-NDFZMnX94cIK2noT3_K:blymx7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/9cecacbeb2b0dced69f227f189054c8953e6db4b/pkg/util/json/random.go#L55)):*
> ```Go
> }
> 
> // WithComplexity returns an option to set maximum complexity of JSON objects..
> ```

nit: extra period.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> > Reviewed 3 of 3 files at r2, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89117)_ status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @miretskiy)
> 
> _[`pkg/util/json/random.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDFZMnX94cIK2noT3_J:-NDFZMnX94cIK2noT3_K:blymx7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/9cecacbeb2b0dced69f227f189054c8953e6db4b/pkg/util/json/random.go#L55)):_
> 
> > ```go
> > }
> > 
> > // WithComplexity returns an option to set maximum complexity of JSON objects..
> > ```
> 
> nit: extra period.

removed

miretskiy commented:
@yuzefovich FYI: Had to push updated version of this code due to some
differences discovered and fixed:
  * Handling of zero prefixed numbers -- this is not allowed , and I needed to add extra handling for this
  * Handling of invalid UTF -- had to update the test
  * Various error messages are different.
Also, it's possible (or even likely) that I'll have to do a vendor change to update json-iter.

yuzefovich commented:
Should I take another look? Let me know when it's ready.

yuzefovich reviewed (approved):
> Reviewed 1 of 3 files at r3, 4 of 6 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89117)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @miretskiy, and @yuzefovich)

___
*[`pkg/util/json/json.go` line 922 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDU8sET9fcgbl1543wD:-NDU8sET9fcgbl1543wE:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf35f1ca864daea74b51c4ee1a0c9f50689931b0/pkg/util/json/json.go#L922)):*
> ```Go
> // readNumber reads json.Number from the iterator and checks to see if
> // json.Number has leading 0's -- which is not allowed.
> // TODO(yevgeniy): jsoniter.Iterator should handle this directly in the it.ReadNumber()
> ```

nit: missing period.
___
*[`pkg/util/json/json.go` line 976 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDU9Tr64_o-rsBI6hsE:-NDU9Tr64_o-rsBI6hsF:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf35f1ca864daea74b51c4ee1a0c9f50689931b0/pkg/util/json/json.go#L976)):*
> ```Go
> 		// Use ReadObjectCB since ReadObject does not distinguish between
> 		// end of object (returned as empty key field), and an empty key
> 		// (i.e. ReadObject cannot handle `{"": ""}`)
> ```

nit: missing period.
___
*[`pkg/util/json/json.go` line 1007 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89117#-NDUA18y9Z76fZCS54Eq:-NDUA18y9Z76fZCS54Er:brkksow) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf35f1ca864daea74b51c4ee1a0c9f50689931b0/pkg/util/json/json.go#L1007)):*
> ```Go
> 	// Errors indicated via it.Error.
> 	if it.Error != nil && it.Error != io.EOF {
> 		err := errors.Handled(it.Error)
> ```

nit: this probably should be extracted in a common helper.


<!-- Sent from Reviewable.io -->

miretskiy commented:
Closing in favor of #89884
